func/internal/podevaluator.go

```diff
@@ -466,35 +473,137 @@ func (pm *podManager) getFuncEvalPodClient(ctx context.Context, image string, tt
+		Name:      "auth-secret",
+		Namespace: pm.namespace,
+	}, secret)
+	if err != nil {
+		if client.IgnoreNotFound(err) != nil {
+			// Error other than "not found" occurred
+			return err
+		}
+		klog.Infof("Secret for private registry pods does not exist and is required.\nGenerating Secret Now")
+		dockerConfigBytes, err := os.ReadFile(registryAuthSecretPath)
+		if err != nil {
+			return err
+		}
+		// Secret does not exist, create it
+		secret = &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "auth-secret",
+				Namespace: pm.namespace,
+			},
+			Data: map[string][]byte{
+				".dockerconfigjson": dockerConfigBytes,
+			},
+			Type: corev1.SecretTypeDockerConfigJson,
+		}
+		err = pm.kubeClient.Create(ctx, secret)
+		if err != nil {
+			return err
+		}
+
+		klog.Infof("Secret created successfully")
+	} else {
+		klog.Infof("Secret already exists")
+	}
+	return nil
+}
+
+// DockerConfig represents the structure of Docker config.json
+type DockerConfig struct {
+	Auths map[string]authn.AuthConfig `json:"auths"`
+}
+
 // imageDigestAndEntrypoint gets the entrypoint of a container image by looking at its metadata.
-func (pm *podManager) imageDigestAndEntrypoint(ctx context.Context, image string) (*digestAndEntrypoint, error) {
+func (pm *podManager) imageDigestAndEntrypoint(ctx context.Context, image string, registryAuthSecretPath string) (*digestAndEntrypoint, error) {
 	start := time.Now()
 	defer func() {
 		klog.Infof("getting image metadata for %v took %v", image, time.Since(start))
 	}()
-	var entrypoint []string
+
 	ref, err := name.ParseReference(image)
 	if err != nil {
+		klog.Errorf("we got an error parsing the ref %v", err)
 		return nil, err
 	}
-	img, err := remote.Image(ref, remote.WithAuthFromKeychain(gcrane.Keychain), remote.WithContext(ctx))
+
+	var auth authn.Authenticator
+	if registryAuthSecretPath != "" && !strings.HasPrefix(image, defaultRegistry) {
+		if err := pm.handleCustomAuth(ctx, registryAuthSecretPath); err != nil {
+			return nil, err
+		}
+
+		auth, err = pm.getCustomAuth(ref, registryAuthSecretPath)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		auth, err = authn.DefaultKeychain.Resolve(ref.Context())
+		if err != nil {
+			klog.Errorf("error resolving default keychain: %v", err)
+			return nil, err
+		}
+	}
+
+	return pm.getImageMetadata(ctx, ref, auth, image)
+}
+
+// handleCustomAuth ensures if images from custom registry's are requested their appropriate credentials are passed onto a secret for fn pods to use when pulling if it doesnt already exist
+func (pm *podManager) handleCustomAuth(ctx context.Context, registryAuthSecretPath string) error {
+	if err := pm.InspectOrCreateSecret(ctx, registryAuthSecretPath); err != nil {
+		return err
+	}
+	return nil
+}
+
+// if a custom image is requested use the secret provided to authenticate
+func (pm *podManager) appendImagePullSecret(image string, registryAuthSecretPath string, podTemplate *corev1.Pod) {
```

Comment:
Consider moving this method down to after the `getBasePodTemplate` method, since it's used immediately after `getBasePodTemplate` in the flow.

Reply:
moved function down as per request

Reply:
Looks good - OK to resolve